### PR TITLE
Implement QueryOps#groupedQuery

### PR DESCRIPTION
**Pull request by yongpingchen**
#236 
- Implement json util method to serialize request object and deserialize response
- Implement QueryOps#groupedQuery to handle grouped query request
- Add small tests

### CODE REVIEW

File: src/internal/JsonUtilities.ts [hunk 1]
```diff
@@ -508,4 +509,52 @@ export function jsonToHistoryState(json: any): HistoryState {
...
+export function jsonToTimeRange(json: any): TimeRange {
+    return new TimeRange(new Date(json["from"]), new Date(json["to"]));
+}
+
+export function jsonToGroupedHistoryStates(json: any): GroupedHistoryStates {
+    let states: Array<HistoryState> = [];
+    let stateJsons = json["objects"];
+    for (var i = 0; i < stateJsons.length; i++) {
+        states.push(jsonToHistoryState(stateJsons[i]));
+    }
```
Comment (shinsuke-nara) [inline src/internal/JsonUtilities.ts]: Can we use `for..of` or `forEach`?

https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html
http://www.buildinsider.net/language/quicktypescript/01

I think using iterator or `forEach` method is easier to understand than using index.

What do you think?

File: src/ops/QueryOps.ts [hunk 2]
```diff
@@ -67,6 +69,47 @@ export class QueryOps extends BaseOp {
...
+                    }
+                    reject(err);
+                })
+            }
+        });
+    }
+
+    groupedQuery(requestObject: QueryGroupedHistoryStatesRequest): Promise<Array<GroupedHistoryStates>> {
+        return new Promise<Array<GroupedHistoryStates>>((resolve, reject) => {
+            if (!requestObject.alias) {
```
Comment (shinsuke-nara) [inline src/ops/QueryOps.ts]: I think `!!` is better than `!` in this case. What do you think?

File: src/ops/QueryOps.ts [hunk 3]
```diff
@@ -67,6 +69,47 @@ export class QueryOps extends BaseOp {
...
+                reject(new ThingIFError(Errors.ArgumentError, "alias is null or empty"));
+            } else if (!KiiUtil.isString(requestObject.alias)) {
+                reject(new ThingIFError(Errors.ArgumentError, "alias is not string"));
+            } else if (!requestObject.range) {
```
Comment (shinsuke-nara) [inline src/ops/QueryOps.ts]: I think !! is better than ! in this case. What do you think?

File: src/ops/QueryOps.ts [hunk 4]
```diff
@@ -67,6 +69,47 @@ export class QueryOps extends BaseOp {
...
+                    headers: this.getHeaders(),
+                    url: this.getUrlOfAlias(requestObject.alias),
+                    body: requestBody,
+                };
+                request(req).then((res) => {
+                    let jsonResults = (<any>res.body)["groupedResults"];
+                    let groupedResults: Array<GroupedHistoryStates> = [];
+
+                    for (var i = 0; i < jsonResults.length; i++) {
+                        let groupedStates = jsonResults[i];
```
Comment (shinsuke-nara) [inline src/ops/QueryOps.ts]: Can we use `for..in` or `forEach`?

File: src/internal/JsonUtilities.ts (hunk 1 above)
Comment (yongpingchen) [inline src/internal/JsonUtilities.ts]: > I think using iterator or forEach method is easier to understand than using index.

Using `for..in` is concise other than understandable. 

However, it is suggested not to use `for...in` for array. 
This the reason: http://stackoverflow.com/questions/500504/why-is-using-for-in-with-array-iteration-a-bad-idea

File: src/ops/QueryOps.ts (hunk 2 above)
Comment (yongpingchen) [inline src/ops/QueryOps.ts]: > I think !! is better than ! in this case. What do you think?

Not, it should be `!`. Reject when `requestObject.alias` is null or undefined.

File: src/internal/JsonUtilities.ts (hunk 1 above)
Comment (shinsuke-nara) [inline src/internal/JsonUtilities.ts]: Thank you for the information:+1:

Comment (yongpingchen) [inline src/internal/JsonUtilities.ts]: Sorry, I made mistake. `for...in` has problem, but`for..of` has not problem. I will use `for...of`

File: src/ops/QueryOps.ts (hunk 2 above)
Comment (shinsuke-nara) [inline src/ops/QueryOps.ts]: Sorry, I misunderstood about `!!`. You are right.

**Pull request merged**